flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/util/DataFormatConverters.java

```diff
@@ -249,7 +251,15 @@ public static DataFormatConverter getConverterForDataType(DataType originDataTyp
 						DataTypes.INT().bridgedTo(Integer.class));
 			case ROW:
 			case STRUCTURED_TYPE:
-				CompositeType compositeType = (CompositeType) fromDataTypeToTypeInfo(dataType);
+				TypeInformation<?> asTypeInfo = fromDataTypeToTypeInfo(dataType);
+				if (asTypeInfo instanceof WrapperTypeInfo && clazz == RowData.class) {
+					LogicalType realLogicalType = ((WrapperTypeInfo<?>) asTypeInfo).toLogicalType();
+					return new RowDataConverter(getFieldCount(realLogicalType));
```

Comment:
In which case will we reach here?

Reply:
All cases that went through `TypeInfoDataTypeConverter#fromDataTypeToTypeInfo` e.g. the legacy sinks.